### ISSUE TITLE
Customize validation metrics during classification training?

### ISSUE THREAD

**Issue by gameveloster** [ultralytics/ultralytics#8908]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

When training a `yolov8x-cls.pt` on a classification task using a train/val/test dataset each containing 4 classes, the validation metrics calculated at the end of each training epoch only shows the class `all`, and only two metrics `top1_acc` and `top5_acc`.

```py
Starting training for 50 epochs...

      Epoch    GPU_mem       loss  Instances       Size
       1/50      6.12G      2.598         12        512: 100%|██████████| 353/353 [01:20<00:00,  4.41it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 67/67 [00:04<00:00, 16.28it/s]
                   all      0.768          1

      Epoch    GPU_mem       loss  Instances       Size
       2/50      6.67G      1.369         12        512: 100%|██████████| 353/353 [01:18<00:00,  4.48it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 67/67 [00:04<00:00, 16.48it/s]
                   all      0.814          1

      Epoch    GPU_mem       loss  Instances       Size
       3/50      6.67G      1.145         12        512: 100%|██████████| 353/353 [01:18<00:00,  4.49it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 67/67 [00:04<00:00, 16.59it/s]
                   all      0.757          1
```

Questions:
1. How do you get it to show the metrics for each of the 4 classes, in addition to `all`?
2. How do you show more metrics like precision, recall, mAP50, mAP50-95?

### Additional

From this code 

https://github.com/ultralytics/ultralytics/blob/014f0b4b8d97b356246673f652ecbfc4e332cec2/ultralytics/models/yolo/classify/val.py#L37-L39

I feel like theres no options to customize the metrics reported at the end of each epoch validation. If so, any good way to overwrite this behavior?

**Comment by glenn-jocher** [comment]
@gameveloster hello! Thanks for reaching out with your questions. 🌟

1. **Class-specific Metrics**: Currently, YOLOv8's classification training primarily reports `top1_acc` and `top5_acc` for the overall dataset (`all`). To obtain metrics for each class individually, you would need to modify the validation code. This involves iterating over class predictions and labels within the validation loop and calculating the desired metrics per class. Unfortunately, as of now, there's no built-in functionality to automatically display per-class metrics during training without custom code modifications.

2. **Additional Metrics (Precision, Recall, mAP)**: For classification tasks, precision, recall, and mAP are not directly reported since they're more commonly associated with object detection tasks. To include these, you'd need to implement additional functions that calculate these metrics based on the model's predictions and the ground truth labels for your dataset. This would also involve custom modifications to the validation code.

Regarding customizing or overriding the behavior to include more metrics, you're correct; the current setup doesn't offer straightforward options for customization through configuration files or command-line arguments. You would need to manually adjust the validation script (`val.py`) to include calculations and logging for the additional metrics you're interested in.

Here's a very basic example of how you might start to modify the code to calculate precision for each class (note: this is a simplified example and would need further development to fully integrate into your workflow):

```python
from sklearn.metrics import precision_score

# Assuming 'preds' and 'targets' are your predictions and labels respectively
# Calculate precision for each class
precisions = precision_score(targets, preds, average=None)

# Log or print precision per class
for i, prec in enumerate(precisions):
    print(f"Class {i} Precision: {prec}")
```

Remember, integrating this into the YOLOv8 validation loop will require careful consideration of how predictions and labels are handled within the loop.

I hope this helps! If you have further questions or need more detailed guidance, feel free to ask. Happy coding! 🚀

**Comment by AEYCstark** [comment]
omg，i have same question！have you solve this problem？
I attempted to improve it in the source code, but its coupling is too strong, and I have no idea where to make changes.QAQ

**Comment by glenn-jocher** [comment]
@AEYCstark hey there! 😊 It seems like you're encountering a similar challenge. Modifying the source for customized metrics might indeed seem daunting due to how the components are intertwined. A good starting point, however, could be the validation script where metrics are computed. For class-specific metrics, you'd typically examine where the model's predictions are processed. Here's a snippet to get you started:

```python
from sklearn.metrics import precision_score

# Example to calculate precision for predictions
# Assume 'preds' are your model predictions, and 'true_labels' are your ground truth labels
precision = precision_score(true_labels, preds, average='micro')  # You can adjust 'average' parameter
print(f"Precision: {precision}")
```

This is a simplified approach and might need adaptation to fit within your specific context. I hope this gives you a bit of direction though. Don't hesitate to reach out if you have more questions. Happy coding! 🚀
